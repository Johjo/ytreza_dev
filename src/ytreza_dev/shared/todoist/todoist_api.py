import json
from dataclasses import dataclass
from typing import Any

import requests

@dataclass
class TodoistProject:
    id: str
    name: str


@dataclass
class TodoistTask:
    name: str
    url: str
    id: str
    project: TodoistProject


class TodoistAPI:
    def __init__(self, api_token: str):
        self._api_token = api_token
        self._base_url = "https://api.todoist.com/rest/v2"

    def get_all_tasks(self) -> list[TodoistTask]:
        tasks = json.loads(self._get_all_tasks_json())
        projects : dict[str, Any] = {p["id"]: p for p in json.loads(self._get_all_projects_json())}
        return [self._to_todoist_task(task, projects) for task in tasks]

    def _get_all_tasks_json(self):
        headers = {
            "Authorization": f"Bearer {self._api_token}"
        }
        response = requests.get(f"{self._base_url}/tasks", headers=headers)
        response.raise_for_status()
        json_content = response.text
        return json_content

    def _get_all_projects_json(self):
        headers = {
            "Authorization": f"Bearer {self._api_token}"
        }
        response = requests.get(f"{self._base_url}/projects", headers=headers)
        response.raise_for_status()
        json_content = response.text
        return json_content

    @staticmethod
    def _to_todoist_task(task, projects: [dict[str, Any]]):
        return TodoistTask(
            name=task["content"],
            url=task["url"],
            id=task["id"],
            project=TodoistProject(id=task["project_id"], name=projects[task["project_id"]]["name"])
        )

    def open_task(self, content: str) -> str:
        headers = {
            "Authorization": f"Bearer {self._api_token}",
            "Content-Type": "application/json"
        }
        response = requests.post(f"{self._base_url}/tasks", headers=headers, data=json.dumps({"content": content}))
        response.raise_for_status()

        json_content = response.text
        task = json.loads(json_content)
        return str(task["id"])

    def close_task(self, task_id: str) -> None:
        headers = {
            "Authorization": f"Bearer {self._api_token}",
            "Content-Type": "application/json"
        }
        response = requests.post(f"{self._base_url}/tasks/{task_id}/close", headers=headers)
        response.raise_for_status()

    def task_by_id(self, task_id: str) -> TodoistTask:
        task = json.loads(self._json_task_by_id(task_id))
        projects: dict[str, Any] = {p["id"]: p for p in json.loads(self._get_all_projects_json())}
        return self._to_todoist_task(task, projects)

    def _json_task_by_id(self, task_id: str) -> str:
        headers = {
            "Authorization": f"Bearer {self._api_token}"
        }
        response = requests.get(f"{self._base_url}/tasks/{task_id}", headers=headers)
        response.raise_for_status()
        json_content = response.text
        return json_content
