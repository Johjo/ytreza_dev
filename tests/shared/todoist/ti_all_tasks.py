import os

from dotenv import load_dotenv
from approvaltests import verify

from ytreza_dev.shared.todoist.todoist_api import TodoistAPI


def test_all_tasks() -> None:
    load_dotenv(dotenv_path=".env.test")
    api_token = os.getenv("TODOIST_API_TOKEN")
    assert api_token, "TODOIST_API_TOKEN not found in .env.test"

    todoist_api = TodoistAPI(api_token)
    tasks = todoist_api.get_all_tasks()
    verify("\n".join([str(task) for task in tasks]), encoding="utf-8")
