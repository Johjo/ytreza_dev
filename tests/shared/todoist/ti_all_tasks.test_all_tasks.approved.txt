TodoistTask(name='Téléchargez des [applications et plugins gratuits supplémentaires](https://todoist.com/fr/downloads?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox) pour 💻,⌚️,🖥,📱 et 📧', url='https://app.todoist.com/app/task/9388991121', id='9388991121', project=TodoistProject(id='2357501003', name='Inbox'), due_date=Option(some=2025-07-28))
TodoistTask(name='Connecter mon calendrier professionnel à Todoist', url='https://app.todoist.com/app/task/9388991491', id='9388991491', project=TodoistProject(id='2357501015', name='Travail 🎯'), due_date=Option(none=None))
TodoistTask(name='Répondez au [questionnaire sur votre méthode de productivité](https://todoist.com/productivity-methods?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox)', url='https://app.todoist.com/app/task/9388991122', id='9388991122', project=TodoistProject(id='2357501003', name='Inbox'), due_date=Option(some=2025-07-29))
TodoistTask(name='Passer en revue ma journée et planifier', url='https://app.todoist.com/app/task/9388991492', id='9388991492', project=TodoistProject(id='2357501015', name='Travail 🎯'), due_date=Option(some=2025-07-29))
TodoistTask(name='Ajouter mes routines _professionnelles_', url='https://app.todoist.com/app/task/9388991493', id='9388991493', project=TodoistProject(id='2357501015', name='Travail 🎯'), due_date=Option(none=None))
TodoistTask(name='buy the milk', url='https://app.todoist.com/app/task/9497217837', id='9497217837', project=TodoistProject(id='2357501003', name='Inbox'), due_date=Option(none=None))
TodoistTask(name='[Faire une pause régulière](https://blog.doist.com/end-work-day/?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work) 🌖', url='https://app.todoist.com/app/task/9388991494', id='9388991494', project=TodoistProject(id='2357501015', name='Travail 🎯'), due_date=Option(none=None))
TodoistTask(name='[Les intégrations de Todoist qui boosteront votre productivité](https://blog.doist.com/todoist-integrations-productivity/?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work)', url='https://app.todoist.com/app/task/9388991495', id='9388991495', project=TodoistProject(id='2357501015', name='Travail 🎯'), due_date=Option(none=None))
TodoistTask(name='Explorer les [modèles professionnels de Todoist](https://todoist.com/fr/templates/category/work?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work)', url='https://app.todoist.com/app/task/9388991496', id='9388991496', project=TodoistProject(id='2357501015', name='Travail 🎯'), due_date=Option(none=None))
